lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -855,6 +855,92 @@ static void optimizeBatchNorm(Function *F) {
   } // For all nodes in the graph.
 }
 
+/// Check if the \p inputs of this method (vector of NodeValues) can be
+/// legal input of CreateConcat function.
+/// \returns the dimension where the Concat will take place. If not
+/// concatenatable, return -1.
+static int getConcatDimensionForInputs(const std::vector<NodeValue>& inputs,
+                                       ConcatNode *origConcatN) {
+  // For the Concat to work, the inputs must only have one
+  // different dim, like <2,1,3>, <2,2,3> can be concatenated
+  // while <2,1,3>, <1,2,3> cannot. So for this function, it tries
+  // to check the inputs fall into the following two cases:
+  //  1). The inputs dims vectors are all the same. e.g. <2,2,3> ... <2,2,3>.
+  //      For this case, the returned dim is the chosen index that can maintain
+  //      the data order as by the original ConcatNode (origConcatN). A example
+  //      is like, the origConcatN concatenates two <4,3> on the 2nd dim, then
+  //      the new Concat can only concatenate on the 1st/2nd dim, as we want the
+  //      concat "granularity" bigger than 3 (which is of the origConcatN).
+  //  2). Only one dim is different for all the dims vectors. e.g. <2,1,3>,
+  //      <2,2,3>, <2,2,3>, <2,3,3>. For this case, the returned dim is this dim  
+  //      i where dims1[i] != dims2[i].
+  auto firstDims = inputs.front().dims();
+  size_t dimSize = firstDims.size();
+  auto origCNInputShape = origConcatN->getInputs().front().dims();
+  int minimalNumOfUnitToLegallyConcat = 1;
+  // Find the minimal number of data units so that we can legally Concat them.
+  for (size_t i = origConcatN->getDim(); i < origCNInputShape.size(); ++i) {
+    minimalNumOfUnitToLegallyConcat *= origCNInputShape[i];
+  }
+
+  ssize_t index = -1;
+  ssize_t preIndex = -1;
+  for (auto &in : inputs) {
```

Comment:
Can you add a comment before this loop to describe what it tries to do?

Reply:
added